onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc

```diff
@@ -89,37 +118,58 @@ bool ReshapeOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputP
 
   Initializer unpacked_tensor(*new_shape_tensor);
   auto new_shape = unpacked_tensor.DataAsSpan<int64_t>();
+
   if (new_shape.empty()) {
     LOGS(logger, VERBOSE) << "New shape of reshape cannot be empty";
     return false;
   }
 
+  // CoreML reshape does not support 0 as a dimension
+  if (allow_zero) {
+    if (std::find(new_shape.begin(), new_shape.end(), int64_t{0}) != new_shape.end()) {
+      LOGS(logger, VERBOSE) << "Reshape does not support new shape with 0 as dimension when allowzero is enabled. "
+                            << "New shape: " << Shape2String(new_shape);
+      return false;
+    }
+  }
+
   std::vector<int64_t> input_shape;
-  if (!GetStaticShape(*input_defs[0], input_shape, logger))
+  if (!input_params.create_mlprogram) {
+    // if we are using NeuralNetwork, input shape must be static
+    if (!GetStaticShape(*input_defs[0], input_shape, logger)) {
+      LOGS(logger, VERBOSE) << "Unable to get shape of input -- input must have static shape for NeuralNetwork.";
+      return false;
+    }
+  }
+
+  // first input must be fixed rank OR (first input has variadic rank AND shape only contains positive integers)
```

Comment:
"OR (first input has variadic rank AND shape only contains positive integers)"

if I understand correctly, we don't handle this, right? might be worth explicitly saying so.